syntax = "proto3";

package eventstore;

service EventStore {
  rpc publish (PublishRequest) returns (PublishResponse) {}
  rpc subscribe (SubscribeRequest) returns (stream Notification) {}
  rpc unsubscribe (UnsubscribeRequest) returns (UnsubscribeResponse) {}
  rpc get (GetRequest) returns (GetResponse) {}
}

message PublishRequest {
  string event_topic = 1;
  string event_info = 2;
}

message PublishResponse {
  string entry_id = 1;
}

message SubscribeRequest {
  string event_topic = 1;
  string group_name = 2;
}

message Notification {
  string event_id = 1;
  double event_ts = 2;
  string event_action = 3;
  string event_data = 4;
}

message UnsubscribeRequest {
  string event_topic = 1;
}

message UnsubscribeResponse {
  bool success = 1;
}

message GetRequest {
  string event_topic = 1;
}

message GetResponse {
  string events = 1;
}
